### PR TITLE
(docs) Rewrite description of the lookup function

### PR DESCRIPTION
**Pull request by nfagerlund**
When this is merged, please resolve DOC-2225.

### CODE REVIEW

File: lib/puppet/functions/lookup.rb [hunk 1]
```diff
@@ -1,153 +1,128 @@
...
+# available), but you can configure it to merge multiple values into one, fail
+# gracefully, and more.
+#
+# When looking up a key, Puppet will search up to three tiers of data, in the
+# following order:
+#
+# 1. Hiera.
+# 2. The current environment's data provider.
+# 3. The indicated module's data provider, if the key is of the form
+#    `<MODULE NAME>::something`.
```
Comment (gguillotte) [inline lib/puppet/functions/lookup.rb]: Can we use a capped bracketed keyword (maybe <KEY>?) in place of "something" for consistency?

File: lib/puppet/functions/lookup.rb [hunk 2]
```diff
@@ -1,153 +1,128 @@
...
+# Arguments in `[square brackets]` are optional.
+#
+# The arguments accepted by `lookup` are as follows:
+#
+# 1. `<NAME>` (string or array) --- The name of the key to look up.
+#     * This can also be an array of keys. If Puppet doesn't find anything for the
+#     first key, it will try again with the subsequent ones, only resorting to a
+#     default value if none of them succeed.
+# 2. `<VALUE TYPE>` (data type) --- A
+# [data type](http://docs.puppetlabs.com/puppet/latest/reference/lang_data_type.html)
```
Comment (gguillotte) [inline lib/puppet/functions/lookup.rb]: s/http/https/

File: lib/puppet/functions/lookup.rb [hunk 3]
```diff
@@ -1,153 +1,128 @@
...
+# found, Puppet will pass the requested key to the lambda and use its result as
+# the default value.
+#
+# #### Merge Behaviors
+#
+# Puppet lookup uses a hierarchy of data sources, and a given key might have
+# values in multiple sources. By default, Puppet returns the first value it finds,
+# but it can also continue searching and merge all the values together.
+#
+# **Note:** Data sources can use the special `lookup_options` metadata key to
```
Comment (gguillotte) [inline lib/puppet/functions/lookup.rb]: Can we indent/blockquote notes here?

File: lib/puppet/functions/lookup.rb (hunk 1 above)
Comment (nfagerlund) [inline lib/puppet/functions/lookup.rb]: Yeah, works for me.

File: lib/puppet/functions/lookup.rb (hunk 2 above)
Comment (nfagerlund) [inline lib/puppet/functions/lookup.rb]: GOOD CATCH

File: lib/puppet/functions/lookup.rb [hunk 4]
```diff
@@ -1,153 +1,128 @@
...
+# request a specific merge behavior for a key. The `lookup` function will use that
+# requested behavior unless you explicitly specify one.
+#
+# The valid merge behaviors are:
+#
+# * `'first'` --- Returns the first value found, with no merging. Puppet lookup's
+# default behavior.
+# * `'unique'` (AKA "array merge") --- Combines any number of arrays and scalar
```
Comment (gguillotte) [inline lib/puppet/functions/lookup.rb]: "AKA" seems awkward here. Can we note that it's an array merge in the description?

File: lib/puppet/functions/lookup.rb [hunk 5]
```diff
@@ -1,153 +1,128 @@
...
+#     values. If Puppet can't find a normal value, it will check this hash for the
+#     requested key before giving up. You can combine this with `default_value` or
+#     a lambda, which will be used if the key isn't present in this hash. Defaults
+#     to an empty hash.
+#     * `'override'` (hash) --- A hash of lookup keys and override values. Puppet
+#     will check for the requested key in the overrides hash _first;_ if found, it
+#     returns that value as the _final_ value, ignoring merge behavior. Defaults
+#     to an empty hash.
+#
+# Finally, `lookup` can take a lambda, which should accept a single parameter.
```
Comment (gguillotte) [inline lib/puppet/functions/lookup.rb]: Why hedge this with "should"? In what circumstances the lambda not take a single parameter?

File: lib/puppet/functions/lookup.rb (hunk 3 above)
Comment (nfagerlund) [inline lib/puppet/functions/lookup.rb]: :+1:

File: lib/puppet/functions/lookup.rb (hunk 4 above)
Comment (nfagerlund) [inline lib/puppet/functions/lookup.rb]: Well... this was one of those arbitrary UI changes, where they changed the name of something. In Hiera, this behavior is called "array", and in Puppet lookup it's called "unique". So I could reword it to say "(called 'array merge' in classic Hiera)" or something.

File: lib/puppet/functions/lookup.rb (hunk 5 above)
Comment (nfagerlund) [inline lib/puppet/functions/lookup.rb]: ah, you're right. I always miss my chance to drop a MUST.

File: lib/puppet/functions/lookup.rb [hunk 6]
```diff
@@ -1,153 +1,128 @@
...
+# values.
+# * `'deep'` --- Combines the keys and values of any number of hashes to return a
+# merged hash. If the same key exists in multiple source hashes, Puppet will
+# recursively merge hash or array values (with duplicate values removed from
+# arrays). For conflicting scalar values, the highest-priority value will win.
+# * `{'strategy' => 'first|unique|hash'}` --- Same as the string versions of these
+# merge behaviors.
+# * `{'strategy' => 'deep', <DEEP OPTION> => <VALUE>, ...}` --- Same as `'deep'`,
+# but can adjust the merge with additional options. The `strategy` key is
+# required, and you can set any of these optional keys:
```
Comment (gguillotte) [inline lib/puppet/functions/lookup.rb]: What does "The `strategy` key is required" mean here? Does it mean you can't set options for the `deep` merge behavior unless you use the `strategy` hash? Or does it mean you always have to use the `strategy` hash? I'm guessing it's the former.

File: lib/puppet/functions/lookup.rb [hunk 7]
```diff
@@ -1,153 +1,128 @@
...
+#     * `'knockout_prefix'` (string or undef) --- A string prefix to indicate a
+#     value should be _removed_ from the final value. Defaults to `undef`, which
```
Comment (gguillotte) [inline lib/puppet/functions/lookup.rb]: removed from the final **result**, instead of the final **value**?

File: lib/puppet/functions/lookup.rb [hunk 8]
```diff
@@ -1,153 +1,128 @@
...
+#
+# @example Look up a key and return the first value found
+#
+#     lookup('ntp::service_name')
+#
+# @example Do a unique merge lookup of class names, then add all of those classes to the catalog (like `hiera_include`)
+#
+#     lookup('classes', Array[String], 'unique').include
+#
+# @example Do a deep hash merge lookup of user data (where the data sources can remove items with a knockout prefix)
```
Comment (gguillotte) [inline lib/puppet/functions/lookup.rb]: Maybe be more specific in the description?

> Do a deep hash merge lookup of user data, but knock out matches prefixed by '--'

**Comment by gguillotte** [comment]
:+1: gonna push the :radio_button:
